cmake: design doc update (#13321)
ARMmbed/mbed-os: docs/design-documents/tools/cmake.md

# CMake Mbed OS

Requirements:
- CMake 3.13 and higher

> - mbed-tools (python 3.6 and higher)

Reviewer: ```suggestion
- `mbed-tools` (python 3.6 and higher)
```

Two steps approach:

- Mbed OS core CMake
- Building an application with mbed-tools

> Definitions and configurations would be defined in CMake files, Mbed app configuration file (`/path/to/app/mbed_app.json`) and Mbed library configuration files (`/path/to/mbed-os/<LIBRARY_NAME>/mbed_lib.json`). The Mbed build system would parse the Mbed configuration files and generate a `CMakeLists.txt` configuration file.

Reviewer: ```suggestion
Definitions and configurations would be defined in CMake files, an Mbed app configuration file (`/path/to/app/mbed_app.json`) and Mbed library configuration files (`/path/to/mbed-os/<LIBRARY_NAME>/mbed_lib.json`). `mbed-tools` would parse the Mbed library and app configuration files and generate an `mbed_config.cmake` file.
```

The following rules must be respected for backward compatibility.

Target labels, components, and features defined in `/path/to/mbed-os/targets/targets.json` are used to help the build system determine which directories contain sources/include files to use in the build process. This is accomplished through a custom CMake function (`mbed_add_cmake_directory_if_labels`) to append a prefix (TARGET_, COMPONENT_ and FEATURE_) and add matching directories to the list of directories to process.

An example, to add `TARGET_STM` in the folder `targets` where we have folders like TARGET_NXP, TARGET_STM, etc:

```
mbed_add_cmake_directory_if_labels("TARGET")
```

If a user selects for example target `NUCLEO_F411RE`, the target defines the label `STM`. As result, the target folder STM is included.

The same could be applied to other labels like features or components.

> To migrate to the new build system, we can provide auto scanning of the module and generate CMake based on what we find or use the way as described above. In both cases, we could stay backward compatible.

Reviewer: I think we need to fix the grammar here, but I'm not sure what we mean by "auto scanning of the module"; do we mean "auto scanning of Mbed OS module and component directories"?

Reviewer 2: File globbing is what I read

Author: I'll remove it, the migration path will come in separate update

## Mbed OS Core (Mbed OS repository)

There are numerous CMake files in the Mbed OS repository tree:

* A `CMakeLists.txt` entry point in the Mbed OS root, describing the top level build specification for the Mbed OS source tree.
* `CMakeLists.txt` entry points in each Mbed OS module subdirectory, describing the build specification for a module or component

A number of CMake scripts are contained in the `mbed-os/cmake` directory:
* `toolchain.cmake` - selects the toolchain script from the `cmake/toolchains` directory, based on the value of the `MBED_TOOLCHAIN` variable
* `profile.cmake` - selects the profile script from the `cmake/profiles` directory, based on the value of the `MBED_PROFILE` variable

> * core.cmake` - selects the core script from the `cmake/cores` directory, based on the value of the `MBED_CPU_CORE` variable

Reviewer: ```suggestion
* `core.cmake` - selects the core script from the `cmake/cores` directory, based on the value of the `MBED_CPU_CORE` variable
```

I missed a backtick here so the formatting was off.

This is not intended to be included by an application.

### 2. Toolchain CMake Scripts

All the toolchain settings are defined in the scripts found in `cmake/toolchains/`.

### 3. Profile CMake Scripts

The build profiles such as release or debug are defined in the scripts found in `cmake/profiles/`.

### 4. MCU Core CMake Scripts

The MCU core definitions are defined in the scripts found in `cmake/cores/`.

### 5. Utilities CMake Scripts

Custom functions/macros used within Mbed OS.

### 6. Application CMake

The CMake script that must be included by all applications using:

```
include(${MBED_ROOT}/cmake/app.cmake)
```

### 7. Component `CMakeLists.txt` Entry Point

This file statically defines the build specification of an Mbed OS component. It contains conditional statements that depend on the configuration parameters generated by `mbed-tools`. The component entry points make use of functions/macros defined in `util.cmake` to conditionally include or exclude directories.
The rule of thumb is to not expose header files that are internal. We would like to avoid having everything in the include paths as we do now.

## Building an Application

> `mbed-tools` is the next generation of command line tooling for Mbed OS. `mbed-tools` replaces the `mbed-cli` and the Python modules in the `mbed-os/tools` directory.

Reviewer 2: ```suggestion
`mbed-tools` is the next generation of command line tooling for Mbed OS. `mbed-tools` replaces `mbed-cli` and the Python modules in the `mbed-os/tools` directory.
```

> `mbed-tools` has consolidated all of the required modules to build Mbed OS, along with the command line interface, into a single Python package which can be installed using standard python packaging tools.

Reviewer 2: ```suggestion
`mbed-tools` consolidates all of the required modules to build Mbed OS, along with the command line interface, into a single Python package which can be installed using standard Python packaging tools.
```

> A user create own CMake file to configure an application, also with `mbed_app.json` configuration file. The building of an app would look like:

Reviewer 2: ```suggestion
Each application contains a top-level CMakeLists.txt file. The `mbed-tools init` command can create this top-level CMakeLists.txt, or a user can create it manually. Each application also has a number of json configuration files. `mbed-tools configure` creates an Mbed configuration CMake file (`.mbedbuild/mbed_config.cmake`). The process for building an application looks like:
```

Reviewer: I would go with "Each application contains a top-level `CMakeLists.txt` entry point" - this is how the CMake documentation describes this source file type https://cmake.org/cmake/help/latest/manual/cmake-language.7.html#directories. It's also consistent with how we refer to them earlier in the design doc.

> A user create own CMake file to configure an application, also with `mbed_app.json` configuration file. The building of an app would look like:

Reviewer 2: I don't think this is accurate anymore. `mbed-tools` creates the Mbed configuration CMake file (`.mbedbuild/mbed_config.cmake`) and we don't expect users to have to edit it manually.

Author: this was about CMakelists.txt in the application, like blinky has in the root directory.

Reviewer: Do we want the tools to create a template top-level `CMakeLists.txt` when the user runs `mbedtools init` (or whatever the command to create a new workspace ends up being called, it's `init` currently)?

Author: that would be ideal, if cmakelists does not exists and they run `init`, yotta had similar init flow as I recall.

I guess we want to avoid scanning folders , so we let user to fill in the template.

1. Get the target configuration
1. Get the Mbed OS configuration (select what modules we need and get their config, paths, etc)
1. Create .mbedbuild/mbed_config.cmake
1. Build an application

### Configuration

The main purpose of `mbed-tools` is to parse the Mbed configuration system's JSON files (`mbed_lib.json`, `mbed_app.json` and `targets.json`). The tool outputs a single CMake configuration script, which is included by `app.cmake` and `mbed-os/CMakeLists.txt`.

To generate the CMake config script (named `mbed_config.cmake`) the user can run the `configure` command:

`mbedtools configure -t <toolchain> -m <target>`

This will output `mbed_config.cmake` in a directory named `.mbedbuild` at the root of the program tree.

`mbed_config.cmake` contains several variable definitions used to select the toolchain, core and profile CMake scripts to be used in the build system generation:
* `MBED_TOOLCHAIN`
* `MBED_TARGET`
* `MBED_CPU_CORE`
* `MBED_PROFILE`

The tools also generate an `MBED_TARGET_LABELS` variable, containing the labels, components and feature definitions from `targets.json`, used to select the required Mbed OS components to be built.

> The macro definitions parsed from the Mbed OS configuration system are also included in `mbed_config.cmake`. The decision was made to remove `mbed_config.h`.

Reviewer 2: ```suggestion
The macro definitions parsed from the Mbed OS configuration system are also included in `mbed_config.cmake`, so it is no longer necessary for the tools to generate any `mbed_config.h` file to define macros.
```

### mbedignore 

> TBD

Reviewer 2: ```suggestion
With a build system that can understand dependencies between applications, features, components, and targets, `mbedignore` is no longer needed. CMake creates a list of exactly what's needed for an application build, rather than do what the legacy tools did and include everything by default, leaving the application to specify what is not needed. It's far more natural to express what you need rather than what you don't need, not to mention more future proof should some new feature appear in Mbed OS which your application would need to ignore.
```

Author: We should think about migration path for users , if there exists a way for us to facilitate migration the old code base - for example, can we find out there is mbedignore and warn users about this functionality is not part of Mbed OS 7 and point them to a migration guide? This should be addressed in mbed-tools (for CMake, mbedignore does not exists). I added the suggestions here as it should be sufficient.

The following rules must be respected for backward compatibility.

We use target labels to include subfolders. A custom CMake function goes through the target labels. If there is a folder in the current CMake source directory matching a target label, it is included. 

An example, to add `TARGET_STM` in the folder `targets` where we have folders like TARGET_NXP, TARGET_STM, etc:

```
mbed_add_cmake_directory_if_labels("TARGET")
```

If a user selects for example target `NUCLEO_F411RE`, the target defines the label `STM`. As result, the target folder STM is included.

The same could be applied to other labels like features or components.

To migrate to the new build system, we can provide auto scanning of the module and generate CMake based on what we find or use the way as described above. In both cases, we could stay backward compatible.

## Mbed OS Core (Mbed OS repository)

There are numerous CMake files in the Mbed OS repository tree:

> * The boiler plate for Mbed OS defined in Mbed OS root (provides way for the build system to overwrite/add configuration)

Reviewer: Do we mean the Mbed OS root `CMakeLists.txt` here? If so I think we should change this to the following:

```suggestion
* A `CMakeLists.txt` entry point in the Mbed OS root, describing the top level build specification for the Mbed OS source tree.
```

* Toolchain settings (toolchain.cmake) - configures the toolchain from cmake/toolchains directory based on the selected toolchain (`MBED_TOOLCHAIN`)
* Profile configuration (profile.cmake) - configures the profile from cmake/profiles directory based on the selected profile (`MBED_PROFILE`)
* Core configuration (core.cmake) - configures the core from cmake/cores directory based on the selected target (`MBED_CPU_CORE`)
* Utilities (util.cmake) - functions or macros we use
* Application cmake (app.cmake) - the application cmake

> * Each component has a CMakeList file (describing the component - what files are to include, target/feature/component selection based on target, etc)

Reviewer: ```suggestion
* `CMakeLists.txt` entry points in each Mbed OS module subdirectory, describing the build specification for a module or component

A number of CMake scripts are contained in the `mbed-os/cmake` directory:
* `toolchain.cmake` - selects the toolchain script from the `cmake/toolchains` directory, based on the value of the `MBED_TOOLCHAIN` variable
* `profile.cmake` - selects the profile script from the `cmake/profiles` directory, based on the value of the `MBED_PROFILE` variable
* core.cmake` - selects the core script from the `cmake/cores` directory, based on the value of the `MBED_CPU_CORE` variable
* `util.cmake` - custom CMake helper functions and macros
* `app.cmake` - contains part of the build specification for an application
```

The next sections will describe static CMake files within Mbed OS Core repository.

> ### 1. Boilerplate CMake

Reviewer: ```suggestion
### 1. Mbed OS `CMakeLists.txt` Entry Point
```

> The main CMakeList file in the Mbed OS repository contains boilerplate required to build Mbed OS. It describes the Mbed OS tree and provides all the options we have in Mbed OS.

Reviewer: ```suggestion
The `CMakeLists.txt` entry point in the root of the Mbed OS repository contains the top level build specification for Mbed OS. This file also includes the auto generated `mbed_config.cmake` script, which is created by `mbed-tools`.
```

This is not intended to be included by an application.

> ### 2. Toolchain CMake

Reviewer: ```suggestion
### 2. Toolchain CMake Scripts
```

All the toolchain settings are defined in the scripts found in `cmake/toolchains/`.

> ### 3. Profile CMake

Reviewer: ```suggestion
### 3. Profile CMake Scripts
```

The build profiles such as release or debug are defined in the scripts found in `cmake/profiles/`.

> ### 4. MCU core CMake

Reviewer: ```suggestion
### 4. MCU Core CMake Scripts
```

The MCU core definitions are defined in the scripts found in `cmake/cores/`.

> ### 5. Utilities CMake

Reviewer: ```suggestion
### 5. Utilities CMake Scripts
```

Custom functions/macros used within Mbed OS.

### 6. Application CMake

The CMake script that must be included by all applications using:

```
include(${MBED_ROOT}/cmake/app.cmake)
```

> ### 7. Component CMake

Reviewer: ```suggestion
### 7. Component `CMakeLists.txt` Entry Point
```

> This file statically defines the structure of an Mbed OS component. It also contains conditional statements that are based on the configuration. Regular CMake expressions and Mbed OS functions/macros are used in it to conditionally include/exclude directories.

Reviewer: ```suggestion
This file statically defines the build specification of an Mbed OS component. It contains conditional statements that depend on the configuration parameters generated by `mbed-tools`. The component entry points make use of functions/macros defined in `util.cmake` to conditionally include or exclude directories.
```

The rule of thumb is to not expose header files that are internal. We would like to avoid having everything in the include paths as we do now.

> ## Building an application

Reviewer: ```suggestion
## Building an Application
```

> mbed-tools is the next generation mbed-cli. It replaces the current mbed-cli and Mbed OS tools (build system part of Mbed OS).

Reviewer: ```suggestion
`mbed-tools` is the next generation of command line tooling for Mbed OS. `mbed-tools` replaces the `mbed-cli` and the Python modules in the `mbed-os/tools` directory.

`mbed-tools` has consolidated all of the required modules to build Mbed OS, along with the command line interface, into a single Python package which can be installed using standard python packaging tools.
```

TBD?:

> - configuration

Reviewer: ```suggestion
### Configuration

The main purpose of `mbed-tools` is to parse the Mbed configuration system's JSON files (`mbed_lib.json`, `mbed_app.json` and `targets.json`). The tool outputs a single CMake configuration script, which is included by `app.cmake` and `mbed-os/CMakeLists.txt`.

To generate the CMake config script (named `mbed_config.cmake`) the user can run the `configure` command:

`mbedtools configure -t <toolchain> -m <target>`

This will output `mbed_config.cmake` in a directory named `.mbedbuild` at the root of the program tree.

`mbed_config.cmake` contains several variable definitions used to select the toolchain, core and profile CMake scripts to be used in the build system generation:
* `MBED_TOOLCHAIN`
* `MBED_TARGET`
* `MBED_CPU_CORE`
* `MBED_PROFILE`

The tools also generate an `MBED_TARGET_LABELS` variable, containing the labels, components and feature definitions from `targets.json`, used to select the required Mbed OS components to be built.

The macro definitions parsed from the Mbed OS configuration system are also included in `mbed_config.cmake`. The decision was made to remove `mbed_config.h`.
```

> - mbedignore

Reviewer: ```suggestion
- mbedignore - TBD
```

> - Application output (what is being generated)

Reviewer: Explanation of the output is in the "Configuration" section above.
```suggestion
```

# CMake Mbed OS

Requirements:
- CMake 3.13 and higher

> - mbed-tools (python 3.x)

Reviewer: ```suggestion
- mbed-tools (python 3.6 and higher)
```

Author: fixed

Two steps approach:

- Mbed OS core CMake
- Application CMake

> All what can be defined should be in CMake files and configuration (app/mbed .json files). Our build system would parse the configuration and generate config CMake file.

Reviewer: ```suggestion
Everything that can be defined should be in CMake files and configuration (mbed_app/mbed_lib.json) files. Our build system would parse the configuration and generate a CMake config file.
```

> Everything that can be defined should be in CMake files and configuration (mbed_app/mbed_lib.json) files. Our build system would parse the configuration and generate a CMake config file.

Reviewer 3: ```suggestion
Definitions and configurations would be defined in CMake files, Mbed app configuration file (`/path/to/app/mbed_app.json`) and Mbed library configuration files (`/path/to/mbed-os/<LIBRARY_NAME>/mbed_lib.json`). The Mbed build system would parse the Mbed configuration files and generate a `CMakeLists.txt` configuration file.
```

> To stay backward compatible, we create common rules as follows.

Reviewer 3: ```suggestion
For backward compatibility, the following rules must be respected:
```

Reviewer: I'd change that to "The following rules must be respected" to avoid duplicating "follow..." in the same sentence.

> We use target labels to include subfolders. The function goes through target labels and if there is a folder in the current CMake source directory, it includes it. 

Reviewer: ```suggestion
We use target labels to include subfolders. A custom CMake function goes through the target labels. If there is a folder in the current CMake source directory matching a target label, it is included. 
```

> We use target labels to include subfolders. A custom CMake function goes through the target labels. If there is a folder in the current CMake source directory matching a target label, it is included. 

Reviewer 3: ```suggestion
Target labels, components, and features defined in `/path/to/mbed-os/targets/targets.json` are used to help the build system determine which directories contain sources/include files to use in the build process. This is accomplished through a custom CMake function (`mbed_add_cmake_directory_if_labels`) to append a prefix (TARGET_, COMPONENT_ and FEATURE_) and add matching directories to the list of directories to process.
```

Reviewer: You don't need the comma before "and features".

Reviewer 3: Oxford Comma: 
https://www.grammarly.com/blog/what-is-the-oxford-comma-and-why-do-people-care-so-much-about-it/

Reviewer: Yes, I know what it is. I was always taught to avoid them :). It's used more in U.S English than British English, where it isn't considered a standard.

Reviewer 3: Tomato tomato...

Reviewer: Let's go with this change, then? I think it's worth adding, even with the fancy extra comma...

An example, to add `TARGET_STM` in the a folder `targets`:

```
mbed_add_cmake_directory_if_labels("TARGET")
```

> The same could be applied to other labels like features or components.

Reviewer 3: ```suggestion
I.e As the `DISCO_L475VG_IOT01A` target inherits the `STM` label and has the `BLE` feature, the `TARGET_STM` directory in `mbed-os/targets/` and the `FEATURE_BLE` directory in `mbed-os/features/` are processed because their parent directories contain `CMakeLists.txt` files that have each respectively `mbed_add_cmake_directory_if_labels("TARGET")` and `mbed_add_cmake_directory_if_labels("FEATURE")`.
```

To migrate to the new build system, we can provide auto scanning of the module and generate CMake based on what we find or use the way as described above. In both cases, we could stay backward compatible.

## Mbed OS Core (Mbed OS repository)

> There are couple of CMakes in the tree.

Reviewer: ```suggestion
There are couple of CMakeList files in the tree.
```

> There are couple of CMakeList files in the tree.

Reviewer 3: ```suggestion
There are numerous CMake Language source files in the Mbed OS repository tree:
```

Reviewer: We should refer to them as CMakeList files, or CMakeLists.txt (as that's the terminology used in CMake and avoids confusion with .cmake file types).

Reviewer 3: This line is talking about a combination of scripts and text files recognised by CMake.
I am reading the CMake Cookbook and I have not seen anywhere where CMake scripts (`*.cmake`) and CMake text files (`CMakeLists.txt`). Where is `CMakeList` mentioned?

Reviewer: In the CMake Wiki they refer to them as "list files", or "listfiles". This is also the commonly used terminology on stackoverflow etc. We should either refer to it as it `CMakeLists.txt` or "CMake list file". I think we should change this to "There are numerous CMakeLists.txt and .cmake files in the Mbed OS repository tree".

Reviewer 3: I like `There are numerous CMakeLists.txt and .cmake files in the Mbed OS repository tree`.

Reviewer 3: They shall be referred as "CMake Language source files" as per the official CMake documentation.

"CMake Language source files in a project are organized into..."

Source: https://cmake.org/cmake/help/latest/manual/cmake-language.7.html

Reviewer: The CMake site seems to be down for me at the moment, but "CMake language source files" sounds reasonable. It's better than just "CMake files" anyway. They also refer to them as "CMake input files" FWIW.

> 1. The boiler plate for Mbed OS defined in Mbed OS root (provides way for the build system to overwrite/add configuration)

Reviewer 3: ```suggestion
* The boilerplate `CMakeLists.txt` file - defined in the root of Mbed OS, it allows the build system to overwrite or append configurations.
```

> 1. Toolchain settings (toolchain.cmake) - configures the toolchain from cmake/toolchains directory based on the selected toolchain (`MBED_TOOLCHAIN`)

Reviewer 3: ```suggestion
* Toolchain settings script (toolchain.cmake) - selects the toolchain CMake script (in `cmake/toolchains/`) to use based on the definition of the `MBED_TOOLCHAIN` variable. This variable is provided by the Mbed tools when creating configuration for the application.
```

> 1. Profile configuration (profile.cmake) - configures the profile from cmake/profiles directory based on the selected profile (`MBED_PROFILE`)

Reviewer 3: ```suggestion
* Profile configuration script (profile.cmake) - selects the build profile CMake script (in `cmake/profiles/`) to use based on the definition of the `MBED_PROFILE` variable. This variable is provided by the user to CMake in the CLI when building the application.
```

> 1. Core configuration (core.cmake) - configures the core from cmake/cores directory based on the selected target (`MBED_CPU_CORE`)

Reviewer 3: ```suggestion
* MCU Core configuration script (core.cmake) - selects the MCU core CMake script (in `cmake/cores/`) to use based on the definition of the `MBED_CPU_CORE` variable. This variable is provided by the Mbed tool after parsing the `targets.json` file.
```

1. Utilities (util.cmake) - functions or macros we use

> 1. Application cmake (app.cmake) - an application includes it to

Reviewer: Seems like something is missing from the end of this sentence?

> 1. Utilities (util.cmake) - functions or macros we use

Reviewer 3: ```suggestion
* Utilities script (util.cmake) - contains custom helper functions or macros.
```

> 1. Application cmake (app.cmake) - the application cmake

Reviewer 3: ```suggestion
* Application script (app.cmake) - contains configurations specific to the generation of Mbed application executable and binary.
```

> 1. Each component has own CMake file (describing the component - what files are to include, target/feature/component selection based on target, etc)

Reviewer: ```suggestion
1. Each component has a CMakeList file (describing the component - what files are to include, target/feature/component selection based on target, etc)
```

> 1. Each component has a CMakeList file (describing the component - what files are to include, target/feature/component selection based on target, etc)

Reviewer 3: ```suggestion
* Mbed OS component specific `CMakeLists.txt` files - Each component has a `CMakeLists.txt` file  that describes which directories/files are to be processed by CMake.
```

The next sections will describe static CMake files within Mbed OS Core repository.

### 1. Boilerplate CMake

> The main CMake file in Mbes OS repository provides just boilerplate for Mbed OS to be built. It described the Mbed OS tree, provides all the options we have in Mbed OS.

Reviewer: ```suggestion
The main CMakeList file in the Mbed OS repository contains boilerplate required to build Mbed OS. It describes the Mbed OS tree and provides all the options we have in Mbed OS.
```

This is not intended to be included by an application.

### 2. Toolchain CMake

> All the toolchain settings are defined in cmake/toolchains cmake files.

Reviewer 3: ```suggestion
All the toolchain settings are defined in the scripts found in `cmake/toolchains/`.
```

### 3. Profile CMake

> The application profiles are defined in cmake/profiles.

Reviewer 3: ```suggestion
The build profiles such as release or debug are defined in the scripts found in `cmake/profiles/`.
```

> ### 4. Core CMake

Reviewer 3: ```suggestion
### 4. MCU Core CMake
```

> The core definitions are defined in cmake/cores.

Reviewer 3: ```suggestion
The microprocessor core definitions are defined in the scripts found in `cmake/cores/`.
```

### 5. Utilities

> Functions or macros that we use within Mbed OS and might be useful for other as well.

Reviewer 3: ```suggestion
Custom functions/macros used within Mbed OS.
```

### 6. Application CMake

> This is the cmake that must be included by an application, using:

Reviewer 3: ```suggestion
The CMake script that must be included by all applications using:
```

```
include(${MBED_ROOT}/cmake/app.cmake)
```

### 7. Component CMake

> This file statically defines the structure of the component within Mbed OS. It is conditionally config based. We use regular CMake expressions plus extend it with own macros to conditionally include/exclude directories. The thumb of the rule, do not expose headers that are internal. We would like to avoid having everything in the include paths as we do now.

Reviewer: ```suggestion
This file statically defines the structure of the component within Mbed OS. It contains conditional statements that are based on the config. We use regular CMake expressions, plus extend it with our own macros, to conditionally include/exclude directories. The rule of thumb: do not expose headers that are internal. We would like to avoid having everything in the include paths as we do now.
```

Author: Thanks, will review all and update.

> This file statically defines the structure of the component within Mbed OS. It contains conditional statements that are based on the config. We use regular CMake expressions, plus extend it with our own macros, to conditionally include/exclude directories. The rule of thumb: do not expose headers that are internal. We would like to avoid having everything in the include paths as we do now.

Reviewer 3: ```suggestion
This file statically defines the structure of an Mbed OS component. It also contains conditional statements that are based on the configuration. Regular CMake expressions and Mbed OS functions/macros are used to conditionally include/exclude directories.
The rule of thumb is to not expose header files that are internal. We would like to avoid having everything in the include paths as we do now.
```

Reviewer: I don't think we need the "in it" here. "Regular CMake expressions and Mbed OS functions/macros are used to conditionally include/exclude directories." Reads better to me.

> `add_subdirectory` always adds the directory to the main CMake.

Reviewer 3: Is it necessary to explain a CMake function here?

Author: I'll remove

```
add_subdirectory(drivers)
```

## Application CMake

> We should provide application CMake functionality with  our own configuration. There are couple of approaches we could take. Statically defined CMake but then this disconnectes config and CMake - as CMake contains configuration for a project (like includes, sources, etc). Our build tool would need to parse CMake to get all paths used in the project or Mbed OS to find out where to look for configuration file. Therefore the build system has a knowledge as it is currently. We use `requires` to include/exclude modules.

Reviewer 3: There are extra spaces in this paragraph.

`mbed-tools` consolidates all of the required modules to build Mbed OS, along with the command line interface, into a single Python package which can be installed using standard Python packaging tools.

> Each application contains a top-level CMakeLists.txt file. The `mbed-tools init` command can create this top-level CMakeLists.txt, or a user can create it manually. Each application also has a number of json configuration files. `mbed-tools configure` creates an Mbed configuration CMake file (`.mbedbuild/mbed_config.cmake`). The process for building an application looks like:

Reviewer 2: ```suggestion
Each application contains a top-level CMakeLists.txt file. The `mbedtools init` command can create this top-level CMakeLists.txt, or a user can create it manually. Each application also has a number of json configuration files. `mbedtools configure` creates an Mbed configuration CMake file (`.mbedbuild/mbed_config.cmake`). The process for building an application looks like:
```

> 2. Parse the application configuration

Reviewer 3: Do you want to carry on using `1` for numbered bullet points so you do not need to remember what the next number is?

Author: yes

Author: will fix
